### DISCUSSION TITLE
Определение наличия ключевого контейнера и документация

### DISCUSSION THREAD

**Discussion by ish-1313** [CryptoPro/libcore#59] [Q&A] [answered]
Всем привет, сейчас ComputeSignature свалилась с "Keyset as registered is invalid"  из за невставленной флешки .
возникла пара вопросов
1. какой наиболее правильный способ определить, что флешка с контейнером вставлена 
2. есть ли где то подробное описание библиотеки, чтобы не задавать дурацких вопросов 
3.
tia 
ish

**Comment by Fasjeit** [comment, marked as answer]
Добрый день.

Для определения наличия контейнера с ключом можно попробовать открыть его по аналогии [с примером](https://dss.cryptopro.ru/libcore/docs/04-%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D1%80%D1%8B/04-05-%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0-%D1%81-%D0%BA%D0%BE%D0%BD%D1%82%D0%B5%D0%B9%D0%BD%D0%B5%D1%80%D0%BE%D0%BC-%D0%BA%D0%BB%D1%8E%D1%87%D0%B0.html#%D0%BE%D1%82%D0%BA%D1%80%D1%8B%D1%82%D0%B8%D0%B5-%D0%BA%D0%BE%D0%BD%D1%82%D0%B5%D0%B9%D0%BD%D0%B5%D1%80%D0%B0-%D0%BA%D0%BB%D1%8E%D1%87%D0%B0) с флагами `CspProviderFlags.NoPrompt | CspProviderFlags.UseExistingKey`:

```csharp
// Используется 80 провайдер и ключ 2012 256, если у вас другой ключ - тип провайдера нужно изменить
using (var provider =
    new Gost3410_2012_256CryptoServiceProvider(
        new CpCspParameters(
        // тип провайдера
        80, 
        // имя провайдера, можно указать пустое и будет использован провайдер типа по умолчанию
        "", 
         // имя контейнера
        "\\\\.\\Aktiv Co. ruToken 0\\b75b5058-19f3-47bf-ad0b-6ad189e28dd4")
        // флаги - без окошек, только существующий ключ (не создавать, если не найден)
        { Flags = CspProviderFlags.NoPrompt | CspProviderFlags.UseExistingKey })) 
{

}
```

Если указанный выше код кинет `System.Security.Cryptography.CryptographicException: 'Набор ключей не существует'` то значит доступа к контейнеру нет и токен с ним скорее всего не вставлен.

Актуальная документация - только та, что указана по ссылке в Readme текущего проекта. Постепенно будем наполнять примерами.

В целом большая часть функциональности совпадает (с точностью до изменения некоторый классов с `{name}` на `Cp{name}`) с КриптоПро.NET, документацию к которому можно найти [тут](https://cpdn.cryptopro.ru/default.asp?url=content/cpnet/html/08bcd27a-1f1c-4494-a996-37d88776309e.htm), и примеры можно посмотреть в его sdk.

**Comment by ish-1313** [comment]
Понял, спасибо. упростить планов нет? :) может wish list какой то есть  ? :)

**Reply by Fasjeit** [reply]
Если речь о документации - как будет позволять время будем улучшать.

**Reply by ish-1313** [reply]
скорее о коде. наверняка где то внутри есть код, проверяющий доступен ли контейнер для данного ключа. поставить public :)

**Reply by Fasjeit** [reply]
Есть, но он находится в коде csp, а не libcore. В рамках данного проекты мы просто перевызываем нужным образом си-шные интерфейсы и обрабатываем их ошибки. В рамках текущего проекта никакие полезные интерфейсы мы не прячем.

**Reply by ish-1313** [reply]
Понятно, спасибо. и мыслей не было, что прячете :)
